ai-platform-unified/tutorials/unofficial/pytorch-image-classification-custom-train-container-gpu-custom-prediction-container-torchserve/README.md

```diff
@@ -0,0 +1,243 @@
+echo "PRED_TASK: ${PRED_TASK}"
+```
+```
+export TUTORIAL_NAME_TRAIN=${ML_FRAMEWORK}-${ML_TASK}-${TRAIN_TASK}
+export TUTORIAL_NAME_PRED=${ML_FRAMEWORK}-${ML_TASK}-${PRED_TASK}
+
+echo "TUTORIAL_NAME: ${TUTORIAL_NAME_TRAIN} for training and ${TUTORIAL_NAME_PRED} for prediction"
+```
+```
+export TUTORIAL_DIR=${PWD}
+
+echo "TUTORIAL_DIR: ${TUTORIAL_DIR}"
+```
+
+### Trainer Task Local Run
+```
+cd ${TUTORIAL_DIR}
+cd trainer && python task.py --epochs 1
+rm -rf data
+rm -rf model
+```
+
+### Custom PyTorch Container for Training
+
+```
+export HOSTNAME=gcr.io
+export IMAGE_TRAIN=${TUTORIAL_NAME_TRAIN}
+export TAG=latest
+export CONTAINER_IMAGE_URI=${HOSTNAME}/${PROJECT}/${IMAGE_TRAIN}:${TAG}
+
+echo "CONTAINER_IMAGE_URI: ${CONTAINER_IMAGE_URI}"
+```
+```
+cd ${TUTORIAL_DIR}
+cd trainer && docker build -t ${CONTAINER_IMAGE_URI} -f Dockerfile .
+```
+```
+docker run --rm ${CONTAINER_IMAGE_URI} --epochs 1
+```
+```
+docker push ${CONTAINER_IMAGE_URI}
+```
+```
+gcloud container images list --repository ${HOSTNAME}/${PROJECT}
+```
+
+### Launch a Custom Training Job
+
+##### Configs
+###### CPU
+```
+export CUST_TRAIN_DISPLAY_NAME_SDK=${TUTORIAL_NAME_TRAIN}-sdk-cpu
+
+export MACHINE_TYPE=n1-standard-4
+export ACCELERATOR_TYPE=ACCELERATOR_TYPE_UNSPECIFIED
+export ACCELERATOR_COUNT=0
+export CONT_ARGS="--batch-size 4 --epochs 5"
+
+```
+
+###### Single GPU
+```
+export CUST_TRAIN_DISPLAY_NAME_SDK=${TUTORIAL_NAME_TRAIN}-sdk-gpu
+
+export MACHINE_TYPE=n1-standard-4
+export ACCELERATOR_TYPE=NVIDIA_TESLA_K80
+export ACCELERATOR_COUNT=1
+export CONT_ARGS="--batch-size 4 --epochs 5"
+```
+
+###### Multiple GPUs
+```
+export CUST_TRAIN_DISPLAY_NAME_SDK=${TUTORIAL_NAME_TRAIN}-sdk-gpus
+
+export MACHINE_TYPE=n1-standard-32
+export ACCELERATOR_TYPE=NVIDIA_TESLA_P100
+export ACCELERATOR_COUNT=4
+export CONT_ARGS="--batch-size 32 --epochs 5"
+```
+
+##### SDK
+```
+echo "CUST_TRAIN_DISPLAY_NAME_SDK: ${CUST_TRAIN_DISPLAY_NAME_SDK}"
+
+echo "MACHINE_TYPE: ${MACHINE_TYPE}"
+echo "ACCELERATOR_COUNT: ${ACCELERATOR_COUNT}"
+echo "ACCELERATOR_TYPE: ${ACCELERATOR_TYPE}"
+echo "CONT_ARGS: ${CONT_ARGS} --epochs 5"
+
+cd ${TUTORIAL_DIR}
+python sdk_custom_container_training.py \
+    --project ${PROJECT} \
+    --bucket ${BUCKET} \
+    --location ${LOCATION} \
+    --display-name ${CUST_TRAIN_DISPLAY_NAME_SDK} \
+    --container-image-uri ${CONTAINER_IMAGE_URI} \
+    --machine-type ${MACHINE_TYPE} \
+    --accelerator-type ${ACCELERATOR_TYPE} \
+    --accelerator-count ${ACCELERATOR_COUNT} \
+    --container-args ${CONT_ARGS}
+```
+###### Output Model Artifact
+```
+gsutil ls gs://${BUCKET}/${CUST_TRAIN_DISPLAY_NAME_SDK}
+```
+
+
+### Model Archive for TorchServe
+
+```
+cd ${TUTORIAL_DIR}
+cd model_server && gsutil cp -r gs://${BUCKET}/${CUST_TRAIN_DISPLAY_NAME_SDK}/model .
+```
+```
+cd ${TUTORIAL_DIR}
+cd model_server && torch-model-archiver \
+     --model-name antandbee \
+     --version 1.0 \
+     --serialized-file ./model/antandbee.pth \
```

Comment:
Nits regarding `antandbee.pth`:

1. Just a note that this file is pretty large and might affect performance for git fetch and push, especially if gets change din the future. See https://docs.github.com/en/github/managing-large-files/working-with-large-files

   It is probably still fine, but I just want to flag in case the team has a best practice for large files like this. For example, would it be better to put it in `gs://cloud-samples-data` instead and have the reader download it as part of these instructions?

2. It seems like "training" and "prediction" parts of this sample are pretty separate: for prediction, the reader uses the provided `antandbee.pth` rather than the one that they trained themselves. This is nice for users who want to skip training, but I guess some readers might prefer the "end to end" workflow where they use their training output directly for prediction.

   Perhaps this is a different case though (and requires some different code), and maybe it's better suited for a Jupyter notebook? So no particular suggestion here, just flagging that this wasn't initially obvious to me; if we work on documentation based on this, we will make sure this gets pointed out to the reader.